File: generated_code/GeminiPro/Java/073_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree that functions as an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again. This implementation includes
 * insert, delete, and search operations. The search operation also performs the splaying
 * action, moving the accessed node to the root of the tree.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     * It holds the integer key and references to its parent, left child, and right child.
     */
    private class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     * @param x The node to be rotated.
     */
    private void rotateLeft(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node x.
     *
     * @param x The node to be rotated.
     */
    private void rotateRight(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.right = x;
            x.parent = y;
        }
    }

    /**
     * Performs the splay operation on the given node, moving it to the root.
     *
     * @param x The node to splay.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rotateRight(parent);
                } else {
                    rotateLeft(parent);
                }
            } else {
                if (x == parent.left && parent == grandParent.left) { // Zig-Zig case (left-left)
                    rotateRight(grandParent);
                    rotateRight(parent);
                } else if (x == parent.right && parent == grandParent.right) { // Zig-Zig case (right-right)
                    rotateLeft(grandParent);
                    rotateLeft(parent);
                } else if (x == parent.left && parent == grandParent.right) { // Zig-Zag case (right-left)
                    rotateRight(parent);
                    rotateLeft(grandParent);
                } else { // Zig-Zag case (left-right)
                    rotateLeft(parent);
                    rotateRight(grandParent);
                }
            }
        }
        this.root = x;
    }

    /**
     * Searches for a key in the tree. If the key is found, the corresponding node
     * is splayed to the root. If not found, the last accessed node is splayed.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }

        Node current = root;
        Node lastNode = root;

        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found, splay the node and return the key
                splay(current);
                return current.key;
            }
        }

        // Key not found, splay the last accessed node
        if (lastNode != null) {
            splay(lastNode);
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree. If the key already exists, the existing
     * node is splayed. If the key is new, it is inserted and the new node is splayed.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;

        // Traverse to find the insertion point or an existing key
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the node and return
                splay(current);
                return;
            }
        }

        // Create and insert the new node
        Node newNode = new Node(key);
        newNode.parent = parent;

        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        // Splay the newly inserted node to the root
        splay(newNode);
    }

    /**
     * Deletes a key from the Splay Tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // Search for the key. This splays the node to the root if found.
        search(key);

        // If the root is null or doesn't have the key, the key was not in the tree.
        if (root == null || root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right subtree becomes the new tree.
            this.root = rightSubtree;
            if (this.root != null) {
                this.root.parent = null;
            }
        } else {
            // Find the maximum node in the left subtree.
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Splay this maximum node to the root of the left subtree.
            // To do this, we temporarily treat the left subtree as the main tree.
            leftSubtree.parent = null; // Disconnect for the splay operation
            this.root = leftSubtree;
            splay(maxNode);

            // After splaying, maxNode is the new root. Its right child is null.
            // We can now attach the original right subtree.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}

File: generated_code/GeminiPro/Python/073_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Splay Tree class implementing a dictionary-like set for integers.
    
    Splay trees are self-balancing binary search trees with the additional property
    that recently accessed elements are quick to access again. It achieves this
    by moving any accessed node to the root of the tree through a series of
    rotations in an operation called "splaying".

    This implementation includes the following methods:
    - insert(key): Adds an integer key to the set.
    - delete(key): Removes an integer key from the set.
    - search(key): Checks for the existence of a key and splays the accessed node.
    """

    class _Node:
        """A private inner class to represent a node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splaying operation on node x, moving it to the root.
        
        The splaying operation consists of a sequence of tree rotations:
        - Zig: If the parent is the root.
        - Zig-Zig: If the node and its parent are both left/right children.
        - Zig-Zag: If the node is a right child and parent is a left child (or vice versa).
        """
        if not x:
            return
            
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (x is left child, parent is right child)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        Performs the splaying operation on the accessed node if the key is found,
        or on the last accessed node (the parent of the potential location) if
        the key is not found.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        last_visited_node = None
        current = self.root
        while current:
            last_visited_node = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # If key was not found, splay the last non-null node visited.
        if last_visited_node:
            self._splay(last_visited_node)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted and then splayed to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Search for the key. This splays the tree, bringing the closest node to the root.
        self.search(key)

        # If key already exists, search would have splayed it and we're done.
        if self.root.key == key:
            return

        # Otherwise, insert the new key and make it the new root.
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
            self.root.parent = new_node
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None
            self.root.parent = new_node
        
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        The parent of the deleted node (or the node that replaced it) is splayed
        to the root.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Search for the key, which splays the node (if found) to the root.
        self.search(key)

        # If key is not in the tree, the root will be the closest node.
        # We only proceed with deletion if the key is actually at the root.
        if self.root.key != key:
            return

        # The node to delete is now the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new root.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # Promote the left subtree to be the new root.
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Join the left and right subtrees.
            # 1. Detach the left subtree.
            left_subtree.parent = None

            # 2. Find the maximum key in the left subtree.
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right

            # 3. Splay this maximum node to the root of the left subtree.
            #    After this, it will have no right child.
            #    We temporarily set self.root to perform splay on the subtree.
            self.root = left_subtree
            self._splay(max_in_left)

            # 4. Attach the original right subtree as the right child of the new root.
            self.root.right = right_subtree
            right_subtree.parent = self.root